### PR TITLE
implement AOP based solution for tracking all API requests

### LINKED ISSUE #1461: implement AOP based solution for tracking all API requests
We need a way to track/audit all API requests coming into xpanse. The reason behind such a requirement is, most cloud providers write all service modification requests to specific systems. In case of Huawei cloud, the data is written to CTS service.

- [ ] Find a way to figure out how to extract for which CSP is the request intended to. Some available options -> make CSP a mandatory URI field in all POST, PUT and DELETE request, or add CSP as header, etc. 
- [ ] Use Spring AOP to `pointCut` all controller POST, PUT and DELETE methods.
- [ ] OrchestratorPlugin interface to have a new method -> auditApiRequest. This method must be to write a simple log line as  default implementation. 
- [ ] Audit only POST, PUT and DELETE requests. 

Please also discuss with @baixinsui about the

### PR DESCRIPTION
**Pull request by WangLiNaruto**
Fixed #1461

image
image
image

image
image

### CODE REVIEW

**Comment by baixinsui** [comment]
@WangLiNaruto  We need to add a test to check if all API methods have annotation @GetCspMethodInfo added. If any API method has not been annotated, the test case will fail and an error message which API method in which class has not been annotated will be thrown.

File: modules/common/src/main/java/org/eclipse/xpanse/common/aop/HttpRequestInfoManager.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import org.eclipse.xpanse.modules.models.common.enums.Csp;
+import org.springframework.web.context.request.RequestAttributes;
+import org.springframework.web.context.request.RequestContextHolder;
+import org.springframework.web.context.request.ServletRequestAttributes;
+
+/**
+ * Bean to provide methods to manage request attributes.
+ */
+@Slf4j
+public class HttpRequestInfoManager {
```
Comment (swaroopar) [inline modules/common/src/main/java/org/eclipse/xpanse/common/aop/HttpRequestInfoManager.java:21]: Is this class really required? We do not seem to use the CSP from http request attribute.

File: modules/api/src/main/java/org/eclipse/xpanse/api/config/GetCspMethodInfo.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Get custom annotations for csp information.
+ */
+@Target(ElementType.METHOD)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface GetCspMethodInfo {
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/config/GetCspMethodInfo.java:19]: This annotation can be renamed as `AuditApiRequest`

File: modules/api/src/main/java/org/eclipse/xpanse/api/config/AuditLogWriter.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    public void controllerMethods() {
+    }
+
+    /**
+     * Validate request parameters.
+     *
+     * @param joinPoint The join point.
+     */
+    @Around("controllerMethods()")
+    public Object setCspIntoHttpRequestInfo(ProceedingJoinPoint joinPoint) throws Throwable {
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/config/AuditLogWriter.java:63]: this method should be called "auditRequest" or something like that.

File: modules/api/src/main/java/org/eclipse/xpanse/api/config/AuditLogWriter.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        String methodName = joinPoint.getSignature().getName();
+        log.info("method: {} start execute.", methodName);
+        Csp csp = getCsp(joinPoint);
+        log.info("Get csp from request attributes: {}", csp);
+        Object result = joinPoint.proceed();
+        if (Objects.isNull(csp)) {
+            return result;
+        }
+        ServletRequestAttributes attributes =
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/config/AuditLogWriter.java:72]: we do not need this extra header.

File: modules/api/src/main/java/org/eclipse/xpanse/api/config/AuditLogWriter.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                csp = getCspFromRequestParams(args);
+            } else {
+                csp = getCspWithMethodInfo(getCspMethodInfo, args);
+            }
+            HttpRequestInfoManager.setCspIntoRequestAttributes(csp);
+        }
+        return csp;
+    }
+
+    private Csp getCspWithMethodInfo(GetCspMethodInfo getCspMethodInfo, Object[] args) {
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/config/AuditLogWriter.java:108]: method name getCspFromMethodInfo

File: modules/models/src/main/java/org/eclipse/xpanse/modules/models/policy/userpolicy/UserPolicyUpdateRequest.java [hunk 6]
```diff
@@ -21,7 +20,6 @@ public class UserPolicyUpdateRequest {
     /**
      * The id of the policy.
      */
-    @NotNull
```
Comment (swaroopar) [inline modules/models/src/main/java/org/eclipse/xpanse/modules/models/policy/userpolicy/UserPolicyUpdateRequest.java:24]: why is this removed?

File: modules/api/src/main/java/org/eclipse/xpanse/api/config/GetCspMethodInfo.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+    /**
+     * The index of params can be used to get csp info with the method.
+     */
+    int[] paramIndexes() default {0};
+
+    /**
+     * Identifies whether csp can be obtained directly.
+     */
+    boolean enabled() default true;
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/config/GetCspMethodInfo.java:45]: two booleans fields. It is confusing. 
My understanding is that the same annotation can be used for all types of methods. So the only boolean flag is necessary to disable audit completely.

File: modules/api/src/main/java/org/eclipse/xpanse/api/controllers/IsvServiceDeployApi.java [hunk 8]
```diff
@@ -81,6 +83,7 @@ public List<DeployedService> listDeployedServicesOfIsv(
     @GetMapping(value = "/services/isv/details/vendor_hosted/{id}",
             produces = MediaType.APPLICATION_JSON_VALUE)
     @ResponseStatus(HttpStatus.OK)
+    @GetCspMethodInfo(isAudit = false)
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/controllers/IsvServiceDeployApi.java:86]: This must be audited.

File: modules/api/src/main/java/org/eclipse/xpanse/api/controllers/OpenTofuMakerWebhookApi.java [hunk 9]
```diff
@@ -49,6 +50,7 @@ public class OpenTofuMakerWebhookApi {
     @PostMapping(value = "${webhook.tofu-maker.deployCallbackUri}/{task_id}", produces =
             MediaType.APPLICATION_JSON_VALUE)
     @ResponseStatus(HttpStatus.OK)
+    @GetCspMethodInfo(enabled = false, methodName = "getCspWithServiceId")
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/controllers/OpenTofuMakerWebhookApi.java:53]: if it is not enabled, why do we need the methodName? Comment applies to few other methods as well.  Please check all.

File: modules/api/src/main/java/org/eclipse/xpanse/api/controllers/ServiceCatalogApi.java [hunk 10]
```diff
@@ -109,6 +111,7 @@ public List<UserOrderableServiceVo> listOrderableServices(
     @GetMapping(value = "/catalog/services/{id}",
             produces = MediaType.APPLICATION_JSON_VALUE)
     @ResponseStatus(HttpStatus.OK)
+    @GetCspMethodInfo(isAudit = false)
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/controllers/ServiceCatalogApi.java:114]: we can audit this.

**Review by swaroopar — CHANGES_REQUESTED** [review]

File: modules/models/src/main/java/org/eclipse/xpanse/modules/models/policy/userpolicy/UserPolicyUpdateRequest.java (hunk 6 above)
Comment (WangLiNaruto) [reply modules/models/src/main/java/org/eclipse/xpanse/modules/models/policy/userpolicy/UserPolicyUpdateRequest.java:24]: image
The API interface has changed and the id has been moved to the path, so this field can be null.

File: modules/api/src/main/java/org/eclipse/xpanse/api/controllers/OpenTofuMakerWebhookApi.java [hunk 11]
```diff
@@ -50,6 +51,7 @@ public class OpenTofuMakerWebhookApi {
     @PostMapping(value = "${webhook.tofu-maker.deployCallbackUri}/{task_id}", produces =
             MediaType.APPLICATION_JSON_VALUE)
     @ResponseStatus(HttpStatus.OK)
+    @AuditApiRequest(enabled = true, methodName = "getCspWithServiceId")
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/controllers/OpenTofuMakerWebhookApi.java:54]: we can disable the audit for all callback methods.

**Review by swaroopar — CHANGES_REQUESTED** [review]

File: modules/api/src/test/java/org/eclipse/xpanse/api/config/XpanseAuditApiRequestTest.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+    public static final String ANNOTATED_METHOD_PACKAGE_NAME = "org.eclipse.xpanse.api.controllers";
+
+    @Test
+    void testAllAnnotatedMethodsHaveAuditApiRequest() {
+        List<Method> annotatedMethods = getAnnotatedMethods(ANNOTATED_METHOD_PACKAGE_NAME);
+
+        for (Method method : annotatedMethods) {
+            assertTrue(method.isAnnotationPresent(AuditApiRequest.class),
+                    "Method " + method.getName() + "  does not have @AuditApiRequest annotation.");
```
Comment (baixinsui) [inline modules/api/src/test/java/org/eclipse/xpanse/api/config/XpanseAuditApiRequestTest.java:25]: It is better to add class name in the error message.

File: modules/api/src/test/java/org/eclipse/xpanse/api/config/XpanseAuditApiRequestTest.java [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        List<Method> annotatedMethods = new ArrayList<>();
+        List<Class<?>> classes = getClasses(packageName);
+        for (Class<?> clazz : classes) {
+            Method[] methods = clazz.getDeclaredMethods();
+            for (Method method : methods) {
+                if (method.isAnnotationPresent(PostMapping.class) ||
+                        method.isAnnotationPresent(GetMapping.class) ||
+                        method.isAnnotationPresent(PutMapping.class) ||
+                        method.isAnnotationPresent(DeleteMapping.class)) {
+                    annotatedMethods.add(method);
```
Comment (baixinsui) [inline modules/api/src/test/java/org/eclipse/xpanse/api/config/XpanseAuditApiRequestTest.java:39]: @RequestMapping and @PatchMapping are missing.

File: modules/api/src/main/java/org/eclipse/xpanse/api/config/GetCspInfoFromRequest.java [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+
+    /**
+     * Get Csp with id of service template.
+     *
+     * @param id id of service template.
+     * @return csp.
+     */
+    public Csp getCspWithTemplateId(String id) {
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/config/GetCspInfoFromRequest.java:78]: getCspWithTemplateId -> getCspFromServiceTemplateId

File: modules/api/src/main/java/org/eclipse/xpanse/api/config/GetCspInfoFromRequest.java [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+        return null;
+    }
+
+    /**
+     * Get Csp with id of service.
+     *
+     * @param id id of service.
+     * @return csp.
+     */
+    public Csp getCspWithServiceId(String id) {
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/config/GetCspInfoFromRequest.java:95]: getCspFromServiceId

File: modules/api/src/main/java/org/eclipse/xpanse/api/config/GetCspInfoFromRequest.java [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+        return null;
+    }
+
+    /**
+     * Get Csp with id of migrate.
+     *
+     * @param id id of service.
+     * @return csp.
+     */
+    public Csp getCspWithMigrateId(String id) {
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/config/GetCspInfoFromRequest.java:112]: getCspFromServiceMigrationId

File: modules/api/src/main/java/org/eclipse/xpanse/api/config/GetCspInfoFromRequest.java [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+        return null;
+    }
+
+    /**
+     * Get Csp with id of user policy.
+     *
+     * @param id id of user policy.
+     * @return csp.
+     */
+    public Csp getCspWithUserPolicyId(String id) {
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/config/GetCspInfoFromRequest.java:134]: getCspFromUserPolicyId

File: modules/api/src/main/java/org/eclipse/xpanse/api/config/GetCspInfoFromRequest.java [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+        return null;
+    }
+
+    /**
+     * Get Csp with taskId of migrate task.
+     *
+     * @param taskId id of migrate task.
+     * @return csp.
+     */
+    public Csp getCspWithWorkflowTaskId(String taskId) {
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/config/GetCspInfoFromRequest.java:168]: getCsFromWorkflowTaskId

File: modules/api/src/main/java/org/eclipse/xpanse/api/config/GetCspInfoFromRequest.java [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+    @Resource
+    private DatabaseServiceMigrationStorage serviceMigrationStorage;
+
+    /**
+     * Get Csp with the URL of Ocl.
+     *
+     * @param url url string of Ocl.
+     * @return csp.
+     */
+    public Csp getCspWithOclLocation(String url) {
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/config/GetCspInfoFromRequest.java:60]: getCspFromOclLocation

File: modules/api/src/main/java/org/eclipse/xpanse/api/config/AuditLogWriter.java [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+
+/**
+ * Writer for audit logs.
+ */
+@Slf4j
+@Aspect
+@Component
+public class AuditLogWriter {
+
+    private static final String DEFAULT_GET_CSP_METHOD_NAME = "getCspFromRequestUri";
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/config/AuditLogWriter.java:42]: what i actually meant is.. we will put this name also in the all the controller methods instead of leaving it blank there. Then it is more clear. We can remove this static parameter from here and simply add this method name to all controller methods where it is empty now.

**Review by swaroopar — CHANGES_REQUESTED** [review]

**Review by swaroopar — APPROVED** [review]
LGTM
